Avoid depending on child process execution when not supported by host OS

topolarity opened the pull request:
In accordance with the requesting issue (#10750):
- `zig test` skips any tests that it cannot spawn, returning success
- `zig run` and `zig build` exit with failure, reporting the command the cannot be run
- `zig clang`, `zig ar`, etc. already punt directly to the appropriate clang/lld main(), even before this change
- Native `libc` Detection is not supported

Additionally, `exec()` and related Builder functions error at run-time, reporting the command that cannot be run

matu3ba reviewed (commented):
just a nit on `can_spawn` else looking good.

Did you check what is still missing?
Any of translate-c, ar, dlltool, lib, ranlib ?

topolarity commented on lib/std/process.zig:
```diff
@@ -936,6 +936,12 @@ pub const can_execv = switch (builtin.os.tag) {
 };
 
+/// Tells whether spawning child processes is supported (e.g. via ChildProcess)
+pub const can_spawn = switch (builtin.os.tag) {
```
Do you mean that `can_spawn` will still need to be updated to be correct for other targets? If so, more than happy to add the TODO

andrewrk replied:
Prefer to make the TODO in a github issue rather than a comment, please :-)

andrewrk commented on lib/std/child_process.zig:
```diff
@@ -126,8 +126,10 @@ pub const ChildProcess = struct {
+        } else if (std.process.can_spawn) {
             return self.spawnPosix();
+        } else {
+            unreachable;
```
I think this spawn function should be untouched, or if you want to make the compile error clearer, put at the top of it:

```zig
if (!std.process.can_spawn) {
    @compileError("the target operating system cannot spawn processes");
}
```

andrewrk commented on build.zig:
```diff
@@ -229,6 +229,11 @@ pub fn build(b: *Builder) !void {
     const version = if (opt_version_string) |version| version else v: {
         const version_string = b.fmt("{d}.{d}.{d}", .{ zig_version.major, zig_version.minor, zig_version.patch });
 
+        // If we can't spawn a process for git, just trust the version string
```
if we can't spawn a process for git, emit a fatal error here saying that it must be provided with `-Dversion-string`.

andrewrk commented on src/Compilation.zig:
```diff
@@ -978,6 +979,9 @@ pub fn create(gpa: Allocator, options: InitOptions) !*Compilation {
             if (!build_options.have_llvm)
                 break :blk false;
 
+            if (!std.process.can_spawn)
+                break :blk false;
```
This should be removed. We can still invoke LLD and Clang even if we cannot spawn processes.

topolarity replied:
Hmm, I had assumed that LLD/Clang couldn't be safely called multiple times, or may never return, due to [a comment in the linker](https://github.com/ziglang/zig/blob/71e0cca7a7957e2f024d2985318e478aa6fb1451/src/link/Elf.zig#L1613-L1615)
> LLD does not behave properly as a library - it calls exit() and does not reset all global data between invocations.

Is that incorrect after all?

andrewrk replied:
It mostly works but it's buggy. I think we should print a warning when invoking LLD for the second time in the same process execution.

andrewrk replied:
Check out 72f6c6e6345030392a3f8cac79862d58359f1e76

topolarity replied:
Great, thanks for the reference. So the preference is to support these calls with warnings, despite the dangers, rather than to block this functionality completely - Makes sense.

I'll go ahead and add non-terminating equivalents of `punt_to_clang` and friends to `src/codegen/llvm.zig`, and update the compiler and linkers to use those on non-spawning OS's

Let me know if you see any issues with that

andrewrk replied:
Sounds mostly good. One suggestion:

Extract the part in `punt_to_clang` that doesn't call exit() into a function called `clangMain`. Make sure the arena memory is freed because we're not going to call exit(). Make *that* function public, leaving `punt_to_clang` private. It can live in main.zig, that's a better place than src/codegen/llvm.zig, which is specifically for lowering AIR to LLVM IR.

andrewrk commented on src/main.zig:
```diff
@@ -2032,6 +2032,9 @@ fn buildOutputType(
         if (opt and cross_target.isDarwin()) {
             fatal("LLD requested with Mach-O object format. Only the self-hosted linker is supported for this target.", .{});
         }
+        if (opt and !std.process.can_spawn) {
+            fatal("LLD requested, but is unsupported on {s} due to inability to spawn child processes. Only the self-hosted linker is supported for this target.", .{@tagName(builtin.os.tag)});
+        }
```
This is incorrect. We can still call LLD without doing it as a child process.

andrewrk commented on src/main.zig:
```diff
@@ -4081,7 +4093,7 @@ extern "c" fn ZigClang_main(argc: c_int, argv: [*:null]?[*:0]u8) c_int;
 
 /// TODO https://github.com/ziglang/zig/issues/3257
-fn punt_to_clang(arena: Allocator, args: []const []const u8) error{OutOfMemory} {
+pub fn punt_to_clang(arena: Allocator, args: []const []const u8) error{OutOfMemory} {
```
I don't think these should be made public or used outside this file

andrewrk reviewed (changes requested):
Thanks! Left you some review comments.

topolarity commented:
Thanks for the review, folks!

> Did you check what is still missing? Any of translate-c, ar, dlltool, lib, ranlib ?

These are already covered without this change :+1: Zig calls directly into `ZigLlvmAr_main`

As for what's left: I thought I had covered all of the usage of `ChildProcess` with my most recent commit, but @andrewrk's request to make referencing `spawn()` a compile-time error (probably a good idea, I agree) is still triggering on an unknown reference I haven't been able to suss out yet.

I'll dive in deeper tomorrow and see if I can eliminate the lingering references to `ChildProcess.spawn`, and address the rest of your comments.

andrewrk commented:
> I'll dive in deeper tomorrow and see if I can eliminate the lingering references to `ChildProcess.spawn`, and address the rest of your comments.

This is a deficiency of zig compile errors; would be nice if it could explain how this dependency is being dragged in. Someday, we'll have it. For now, a workaround is to temporarily delete the spawn function from the std lib and then it tends to generate more helpful errors.

topolarity commented:
> This is a deficiency of zig compile errors; would be nice if it could explain how this dependency is being dragged in. Someday, we'll have it. For now, a workaround is to temporarily delete the spawn function from the std lib and then it tends to generate more helpful errors.

This tip was a huge help; ended up being short work to track things down.

I believe I've addressed all the feedback in the latest changes now. Let me know if anything still needs work :+1:

zigazeljko commented:
Why is `argvCmd` needed? Doesn't `std.mem.join` do exactly the same?

topolarity commented:
> Why is `argvCmd` needed? Doesn't `std.mem.join` do exactly the same?

Nice catch. I updated the PR to use `std.mem.join`

andrewrk commented:
Excellent work. Congrats on landing your first Zig PR :tada:
